framework/src/functions/PiecewiseConstant.C

```diff
@@ -17,76 +17,49 @@ InputParameters
 {
   InputParameters params = PiecewiseTabularBase::validParams();
-  MooseEnum direction("left right", "left");
+  MooseEnum direction("LEFT RIGHT CENTERED", "LEFT");
   params.addParam<MooseEnum>(
       "direction", direction, "Direction to look to find value: " + direction.getRawNames());
   params.addClassDescription("Defines data using a set of x-y data pairs");
   return params;
 }
 
-PiecewiseConstant::DirectionEnum
-PiecewiseConstant::getDirection(const std::string & direction)
-{
-  DirectionEnum dir = PiecewiseConstant::UNDEFINED;
-  if (direction == "left")
-  {
-    dir = PiecewiseConstant::LEFT;
-  }
-  else if (direction == "right")
-  {
-    dir = PiecewiseConstant::RIGHT;
-  }
-  else
-  {
-    mooseError("Unknown direction in PiecewiseConstant");
-  }
-  return dir;
-}
-
 PiecewiseConstant::PiecewiseConstant(const InputParameters & parameters)
-  : PiecewiseTabularBase(parameters), _direction(getDirection(getParam<MooseEnum>("direction")))
+  : PiecewiseTabularBase(parameters),
+    _direction(getParam<MooseEnum>("direction").template getEnum<DirectionEnum>())
 {
 }
 
 Real
 PiecewiseConstant::value(Real t, const Point & p) const
 {
-  Real func_value = 0.0;
   const Real x = _has_axis ? p(_axis) : t;
 
   unsigned i = 1;
   const unsigned len = functionSize();
-  const Real toler = 1e-14;
+  const Real tolerance = 1e-14;
 
   // endpoint cases
-  if ((_direction == LEFT && x < (1 + toler) * domain(0)) ||
-      (_direction == RIGHT && x < (1 - toler) * domain(0)))
-  {
-    func_value = range(0);
-    i = len;
-  }
-  if ((_direction == LEFT && x > (1 + toler) * domain(len - 1)) ||
-      (_direction == RIGHT && x > (1 - toler) * domain(len - 1)))
-  {
-    func_value = range(len - 1);
-    i = len;
-  }
+  if ((_direction == DirectionEnum::LEFT && x < (1 + tolerance) * domain(0)) ||
+      (_direction == DirectionEnum::RIGHT && x < (1 - tolerance) * domain(0)) ||
+      (_direction == DirectionEnum::CENTERED && x < (1 + tolerance) * domain(0)))
+    return _scale_factor * range(0);
+  else if ((_direction == DirectionEnum::LEFT && x > (1 + tolerance) * domain(len - 1)) ||
+           (_direction == DirectionEnum::RIGHT && x > (1 - tolerance) * domain(len - 1)) ||
+           (_direction == DirectionEnum::CENTERED && x > (1 - tolerance) * domain(len - 1)))
+    return _scale_factor * range(len - 1);
 
   for (; i < len; ++i)
   {
-    if (_direction == LEFT && x < (1 + toler) * domain(i))
-    {
-      func_value = range(i - 1);
-      break;
-    }
-    else if ((_direction == RIGHT && x < (1 - toler) * domain(i)))
-    {
-      func_value = range(i);
-      break;
-    }
+    if (_direction == DirectionEnum::LEFT && x < (1 + tolerance) * domain(i))
+      return _scale_factor * range(i - 1);
+    else if ((_direction == DirectionEnum::RIGHT && x < (1 - tolerance) * domain(i)))
+      return _scale_factor * range(i);
+    else if ((_direction == DirectionEnum::CENTERED && x < (1 - tolerance) * domain(i)))
```

Comment:
`(1-tolerance)*domain(i)` shifts a positive `domain(i)` to the left, but it shifts a negative `domain(i)` to the _right_. That seems problematic to me.

Reply:
Oh, not a bad point, but that's a problem with all of these classes, not just this new method, I'll fix that up.

Reply:
on the negative x axis this largely works, but tolerance should have the opposite sign!